libgnucash/engine/gnc-datetime.cpp

```diff
@@ -541,54 +541,62 @@ GncDateTimeImpl::timestamp()
     return str.substr(0, 8) + str.substr(9, 15);
 }
 
+using DateFormatPtr = std::shared_ptr<icu::DateFormat>;
+using CalendarPtr = std::shared_ptr<icu::Calendar>;
+
+static std::tuple<DateFormatPtr, CalendarPtr>
+locale_to_formatter_and_calendar (const std::string locale_str)
+{
+    static std::map<std::string, std::optional<std::tuple<DateFormatPtr, CalendarPtr>>> cache;
+    auto& tuple = cache[locale_str];
+    if (!tuple)
+    {
+        auto locale = icu::Locale::createCanonical (locale_str.c_str());
+        std::shared_ptr<icu::DateFormat> formatter(icu::DateFormat::createDateInstance(icu::DateFormat::kDefault, locale));
+        if (formatter == nullptr)
+            throw std::invalid_argument ("Cannot parse string");
+
+        UErrorCode status = U_ZERO_ERROR;
+        std::shared_ptr<icu::Calendar> calendar(icu::Calendar::createInstance(locale, status));
+        if (U_FAILURE(status))
+            throw std::invalid_argument ("Cannot parse string");
+
+        tuple = std::make_tuple<DateFormatPtr, CalendarPtr>(std::move(formatter), std::move(calendar));
```

Comment:
You forgot to add the tuple to the cache.

Reply:
I didn't: see the magic & in `auto&` ☺️
I did test via printfs

Reply:
The std::move is slightly beyond my current understanding but it's a requirement for compilation to succeed. Chatgpt did suggest it.

Reply:
> see the magic & in auto&

Very clever. I did miss that indeed.

> The std::move is slightly beyond my current understanding 

```
[std::tuple](http://en.cppreference.com/w/cpp/utility/tuple)<unwrap_decay_t<Types>...> make_tuple(Types&&... args)
```
In template context `Types&&` means [forwarding reference](https://en.cppreference.com/w/cpp/language/reference#Forwarding_references). It needs either a reference or an lvalue. A named variable is an rvalue so it needs to be cast into an lvalue. `std::move` does that. `std::move` is apparently confusing. Maybe the committee should have called it `lvalue_cast` instead.